### ISSUE TITLE
[Python][macOS] 3.10 wheel is mislabeled as universal2

### ISSUE THREAD

**Issue by DriveAthelete** [grpc/grpc#28387]
```
(venv) grigoryvorobiev@MacBook-Pro-Grigory pythonProject1 % python main.py                    
Traceback (most recent call last):
  File "/Users/grigoryvorobiev/PycharmProjects/pythonProject1/main.py", line 1, in <module>
    from google.cloud import translate
  File "/Users/grigoryvorobiev/PycharmProjects/pythonProject1/venv/lib/python3.10/site-packages/google/cloud/translate/__init__.py", line 17, in <module>
    from google.cloud.translate_v3.services.translation_service.client import (
  File "/Users/grigoryvorobiev/PycharmProjects/pythonProject1/venv/lib/python3.10/site-packages/google/cloud/translate_v3/__init__.py", line 17, in <module>
    from .services.translation_service import TranslationServiceClient
  File "/Users/grigoryvorobiev/PycharmProjects/pythonProject1/venv/lib/python3.10/site-packages/google/cloud/translate_v3/services/translation_service/__init__.py", line 16, in <module>
    from .client import TranslationServiceClient
  File "/Users/grigoryvorobiev/PycharmProjects/pythonProject1/venv/lib/python3.10/site-packages/google/cloud/translate_v3/services/translation_service/client.py", line 25, in <module>
    from google.api_core import gapic_v1  # type: ignore
  File "/Users/grigoryvorobiev/PycharmProjects/pythonProject1/venv/lib/python3.10/site-packages/google/api_core/gapic_v1/__init__.py", line 16, in <module>
    from google.api_core.gapic_v1 import config
  File "/Users/grigoryvorobiev/PycharmProjects/pythonProject1/venv/lib/python3.10/site-packages/google/api_core/gapic_v1/config.py", line 23, in <module>
    import grpc
  File "/Users/grigoryvorobiev/PycharmProjects/pythonProject1/venv/lib/python3.10/site-packages/grpc/__init__.py", line 22, in <module>
    from grpc import _compression
  File "/Users/grigoryvorobiev/PycharmProjects/pythonProject1/venv/lib/python3.10/site-packages/grpc/_compression.py", line 15, in <module>
    from grpc._cython import cygrpc
ImportError: dlopen(/Users/grigoryvorobiev/PycharmProjects/pythonProject1/venv/lib/python3.10/site-packages/grpc/_cython/cygrpc.cpython-310-darwin.so, 0x0002): tried: '/Users/grigoryvorobiev/PycharmProjects/pythonProject1/venv/lib/python3.10/site-packages/grpc/_cython/cygrpc.cpython-310-darwin.so' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64e')), '/usr/lib/cygrpc.cpython-310-darwin.so' (no such file)
```
The following problem: on 3.9 python everything works, but when upgrading to 3.10, the following problem an import error occurs
Are there any solutions?

**Comment by lidizheng** [comment]
According to the error details, it seems you own a Macbook with M1 chip, we currently don't have the machine to build binary wheels for them. This error is not Python version specific. On the other hand, you can try to emulate x86 architecture:

```sh
arch -x86_64 /bin/bash
# Then install Python, Python packages, and run the application
```

**Comment by benesch** [comment]
> This error is not Python version specific.

Actually, it is. I similarly am able to install `grpcio` on Python 3.9 but not on Python 3.10, just like the issue reporter.

I agree that you wouldn't _expect_ this to depend on the Python version, and indeed that gives us a clue as to what the problem is. On Python 3.9, `pip install grpcio` builds from source (and succeeds), while on Python 3.10, `pip install grpcio` downloads a precompiled wheel... for the wrong architecture. And indeed... grpcio is publishing this wheel: [grpcio-1.43.0-cp310-cp310-macosx_10_10_universal2.whl ](https://files.pythonhosted.org/packages/6f/58/1850083d79702330bf94d44ec14404a60f1d5fcd2b3697b6bd78ba3d998f/grpcio-1.43.0-cp310-cp310-macosx_10_10_universal2.whl).

Note the "universal2", which is incorrectly claiming that the wheel supports both x86_64 and arm64 architectures.

Not having an M1 mac is fine, but please adjust your build process to tag the wheel with x86_64 instead of universal2 so that M1 macs don't incorrectly download the wheel.

**Comment by davissclark** [comment]
> > This error is not Python version specific.
> 
> Actually, it is. I similarly am able to install `grpcio` on Python 3.9 but not on Python 3.10, just like the issue reporter.
> 
> I agree that you wouldn't _expect_ this to depend on the Python version, and indeed that gives us a clue as to what the problem is. On Python 3.9, `pip install grpcio` builds from source (and succeeds), while on Python 3.10, `pip install grpcio` downloads a precompiled wheel... for the wrong architecture. And indeed... grpcio is publishing this wheel: [grpcio-1.43.0-cp310-cp310-macosx_10_10_universal2.whl ](https://files.pythonhosted.org/packages/6f/58/1850083d79702330bf94d44ec14404a60f1d5fcd2b3697b6bd78ba3d998f/grpcio-1.43.0-cp310-cp310-macosx_10_10_universal2.whl).
> 
> Note the "universal2", which is incorrectly claiming that the wheel supports both x86_64 and arm64 architectures.
> 
> Not having an M1 mac is fine, but please adjust your build process to tag the wheel with x86_64 instead of universal2 so that M1 macs don't incorrectly download the wheel.

Super annoying.

**Comment by lidizheng** [comment]
Thanks for the report. We are using the same build process for macOS wheels right now. And we uses `auditwheel` to make sure we are labelling the wheels correctly. It seems this tagging behavior is changed upstream for 3.10.

We might need some volunteer here to investigate why `setuptools` or `auditwheel` is tagging the wheel as `universal2`.

**Comment by benesch** [comment]
Doesn't auditwheel only work for Linux?

**Comment by lidizheng** [comment]
Whoops, yep, I was wrong about the `auditwheel` part. It's intended for manylinux: https://github.com/grpc/grpc/blob/7cfd399dd6f405929da89b3d8124e204b4891513/tools/run_tests/artifacts/artifact_targets.py#L150

**Comment by gabriel-viviani** [comment]
I'm running on the same problem from using mac air m1 and I've tried multiple python versions, including isolating python with miniconda.
Although when I try to use older versions (3.9.9) to install grpcio I only can see:
![image](https://user-images.githubusercontent.com/19916771/149577131-80ee6d79-723f-4cae-bf9c-32ded07d905f.png)

**Comment by lidizheng** [comment]
The log suggests that pyenv picked Python 3.10.1.

**Comment by amitsaha** [comment]
> We might need some volunteer here to investigate why setuptools or auditwheel is tagging the wheel as universal2.

I am happy to do the investigation i will post some notes when i get started.

**Comment by amitsaha** [comment]
@lidizheng if you could point me to the instructions to build and tag the wheel and hence investigate it on M1, i might be able to give you the information you are looking for.

**Comment by lidizheng** [comment]
@amitsaha All Python wheels are built by this script for Linux/macOS: https://github.com/grpc/grpc/blob/master/tools/run_tests/helper_scripts/build_python.sh

(Sorry for its messiness)

**Comment by SaneBow** [comment]
Encountered the same error here with Python 3.10 on Macbook M1.
Quick solution for users is to force pip to build the wheel: `pip install --no-binary :all: grpcio --ignore-installed`

**Comment by amitsaha** [comment]
@lidizheng could we update the CI scripts to use this new option added by `cibuildwheel`? https://github.com/pypa/cibuildwheel/blob/461bed51d3619c4c670f008148de8ce1465ba586/docs/faq.md#building-macos-wheels-for-apple-silicon--apple-silicon

**Comment by amitsaha** [comment]
> @amitsaha All Python wheels are built by this script for Linux/macOS: https://github.com/grpc/grpc/blob/master/tools/run_tests/helper_scripts/build_python.sh
> 
> (Sorry for its messiness)

OK, I have (after some manual hand wrangling to keep the built wheels):

```

# On Mac M1
$ git clone <grpc>
$ GRPC_PYTHON_BUILD_SYSTEM_ZLIB=1 PYTHON=python3 ./tools/run_tests/helper_scripts/build_python.sh
```

The above command is ran inside of a Python 3.10 virtual environment.

I modified the locally installed `pip` in the virtual environment to not delete the built wheels.

I have the following wheels built now:

```
grpcio_health_checking-1.44.0.dev0-py3-none-any.whl
grpcio_admin-1.44.0.dev0-py3-none-any.whl
grpcio_csds-1.44.0.dev0-py3-none-any.whl
grpcio_tools-1.44.0.dev0-cp310-cp310-macosx_10_10_universal2.whl
grpcio_channelz-1.44.0.dev0-py3-none-any.whl
grpcio_testing-1.44.0.dev0-py3-none-any.whl
grpcio-1.44.0.dev0-cp310-cp310-macosx_10_10_universal2.whl
grpcio_status-1.44.0.dev0-py3-none-any.whl
grpcio_reflection-1.44.0.dev0-py3-none-any.whl
grpcio_tests-1.44.0.dev0-py3-none-any.whl
```
Now, i use them to build an example protobuf file:

```
$ cat example.proto 
syntax = "proto3";

// The greeting service definition.
service Greeter {
  // Sends a greeting
  rpc SayHello (HelloRequest) returns (HelloReply) {}
}

// The request message containing the user's name.
message HelloRequest {
  string name = 1;
}

// The response message containing the greetings
message HelloReply {
  string message = 1;
}


$ python -m grpc_tools.protoc --proto_path . --python_out=. --grpc_python_out=. example.proto
```

Everything works. So, I suppose our goal for this bug is to do the same above and check on MacOS intel, why the wheels are being tagged as `universal`.

**Comment by lidizheng** [comment]
@amitsaha Can you try to run the helloworld example from the universal wheel you built? The `.so` might cause issue when it's loaded. https://github.com/grpc/grpc/tree/master/examples/python/helloworld

**Comment by benesch** [comment]
This issue looks relevant: https://github.com/pypa/wheel/issues/406

In particular:

> is your CPython a universal2 one ?
> If so, the default platform tag will always be universal2 when building a platform specific wheel.

https://github.com/pypa/wheel/issues/406#issuecomment-848095295

It seems like maybe `ARCHFLAGS` just needs to be set correctly?

**Comment by lidizheng** [comment]
@benesch Thank you! I tried to add this label, let's see if it fixes the wheel https://github.com/grpc/grpc/pull/28714.

**Comment by xinluleo** [comment]
> Encountered the same error here with Python 3.10 on Macbook M1. Quick solution for users is to force pip to build the wheel: `pip install --no-binary :all: grpcio --ignore-installed`

Thanks. This enables me to resolve the previous error message. However, I am seeing the following error now:
```
File "/Users/xin/Program/Pinecone/iac/envaas/./dev.py", line 7, in <module>
    from pulumi import automation as pa
  File "/Users/xin/miniconda3/envs/iac_py3/lib/python3.10/site-packages/pulumi/__init__.py", line 33, in <module>
    from .config import (
  File "/Users/xin/miniconda3/envs/iac_py3/lib/python3.10/site-packages/pulumi/config.py", line 22, in <module>
    from . import log
  File "/Users/xin/miniconda3/envs/iac_py3/lib/python3.10/site-packages/pulumi/log.py", line 22, in <module>
    from .runtime.settings import get_engine
  File "/Users/xin/miniconda3/envs/iac_py3/lib/python3.10/site-packages/pulumi/runtime/__init__.py", line 29, in <module>
    from .mocks import (
  File "/Users/xin/miniconda3/envs/iac_py3/lib/python3.10/site-packages/pulumi/runtime/mocks.py", line 24, in <module>
    from . import rpc, rpc_manager
  File "/Users/xin/miniconda3/envs/iac_py3/lib/python3.10/site-packages/pulumi/runtime/rpc.py", line 29, in <module>
    from . import known_types, settings
  File "/Users/xin/miniconda3/envs/iac_py3/lib/python3.10/site-packages/pulumi/runtime/settings.py", line 22, in <module>
    import grpc
  File "/Users/xin/miniconda3/envs/iac_py3/lib/python3.10/site-packages/grpc/__init__.py", line 22, in <module>
    from grpc import _compression
  File "/Users/xin/miniconda3/envs/iac_py3/lib/python3.10/site-packages/grpc/_compression.py", line 15, in <module>
    from grpc._cython import cygrpc
ImportError: dlopen(/Users/xin/miniconda3/envs/iac_py3/lib/python3.10/site-packages/grpc/_cython/cygrpc.cpython-310-darwin.so, 0x0002): symbol not found in flat namespace '_CFRelease'
```

Tried with both 1.43.0 and 1.41.1 version grpc, got the same above error.

**Comment by lidizheng** [comment]
The explicit `ARCHFLAGS` failed to change the labelling of macOS wheels for Python 3.10: https://github.com/grpc/grpc/pull/28714

We might need further investigation.

**Comment by pboulos** [comment]
> Encountered the same error here with Python 3.10 on Macbook M1. Quick solution for users is to force pip to build the wheel: `pip install --no-binary :all: grpcio --ignore-installed`

Hello all. Unfortunately this isn't solving it for me. The behaviour I'm experiencing is that it installs successfully using the above method, but upon compiling the gRPC code I get:

`ImportError: dlopen(/Users/peter/.virtualenvs/webdev-testing/lib/python3.10/site-packages/grpc_tools/_protoc_compiler.cpython-310-darwin.so, 0x0002): tried: '/Users/peter/.virtualenvs/webdev-testing/lib/python3.10/site-packages/grpc_tools/_protoc_compiler.cpython-310-darwin.so' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64e')), '/usr/local/lib/_protoc_compiler.cpython-310-darwin.so' (no such file), '/usr/lib/_protoc_compiler.cpython-310-darwin.so' (no such file)`

Looking for suggestions as I feel as though I've exhausted all options.

**Comment by Laurens-T** [comment]
@pboulos While this is not entirely related to this issue, I've had lots of issues with the protoc compiler myself on an M1 mac. 

In the end, I ended up installing `protoc` with Brew and following the steps as explained here: https://github.com/protocolbuffers/protobuf/issues/8820#issuecomment-961552604

**Comment by MURSHIG** [comment]
I did preinstall grpc and grpc-tools with some pip magic:
pip install --no-binary :all: grpcio --ignore-installed
and
pip install --no-binary :all: grpcio-tools --ignore-installed

**Comment by MURSHIG** [comment]
> This answer works for me, thank you @MURSHIG

Welcome

**Comment by parisnakitakejser** [comment]
> I did preinstall grpc and grpc-tools with some pip magic: pip install --no-binary :all: grpcio --ignore-installed and pip install --no-binary :all: grpcio-tools --ignore-installed

Thanks a lot! so long time on 2 simple lines for python 3.10 on M1 🥇

**Comment by GlenDC** [comment]
Are there plans to fix this without having to add these configs?
I am running on a non M1 Macbook by the way (just a good old intel) and I never the less have these issues once running Python 3.10.

Would love to see this fixed.

**Comment by pietrodn** [comment]
It would be nice to have a new patch version released without the defective wheel so that Apple Silicon users can automatically download the source distribution instead.

Unfortunately, [it isn't possible to delete individual wheels](https://snarky.ca/what-to-do-when-you-botch-a-release-on-pypi/) from a PyPI release, but only to replace them.

**Comment by lclpedro** [comment]
if doesn't resolved above, try install python 3.9.8 or superior..

**Comment by RonquilloAeon** [comment]
For anyone using nox, this worked: `session.install("--no-binary", ":all:", "grpcio", "--ignore-installed")`

**Comment by hughhan1** [comment]
Does anyone happen to know what the fix would be if using `pipenv` with a `Pipfile`? I don't explicitly depend on `grpcio`, but I have many dependencies that need `grpcio` in the dependency graph.

I tried the following, but with no success.
```
PIP_NO_BINARY=grpcio pipenv install
```

**Comment by Tenzer** [comment]
With `pipenv` (and Poetry) you can launch a shell in the virtual environment with `pipenv shell` and then run the manual `pip` commands to work around this issue. The tools won't try to install the package again if the installed version matches what it expects.

**Comment by pietrodn** [comment]
The problem is still present with 1.46.0, as another incorrectly labeled wheel was released on PyPI (`grpcio-1.46.0-cp310-cp310-macosx_10_10_universal2.whl`)

**Comment by ajatkj** [comment]
If you're using poetry with python3.10 on M1, you can (replace grpcio-tools with whichever package you're trying to install)
1. remove the dependency using `poetry remove grpcio-tools`
2. run `poetry run pip install --no-binary :all: grpcio-tools --ignore-installed
3. run `poetry add --lock grpcio-tools` to update the lock file

Step 2 will build the dependency from source. This is the only option that seems to be working as of now.

**Comment by joshfriend** [comment]
Poetry has added support for `installer.no-binary`: https://github.com/python-poetry/poetry/pull/5609

**Comment by fredrikaverpil** [comment]
> Poetry has added support for `installer.no-binary`: https://github.com/python-poetry/poetry/pull/5609

That's in Poetry 1.2, slated for release candidate in a few weeks.

You'll have to use Poetry from the main branch of you want to use the feature now. Example pipx installation:

```
pipx install --suffix=@master --force git+https://github.com/python-poetry/poetry.git

poetry@master add --no-binary grpcio
```

**Comment by tamerkadan1988** [comment]
am having the same issue over here : 
https://github.com/googleapis/python-bigquery/issues/1262

**Comment by yehormasalitin** [comment]
Any updates on this issue?
Thank you

**Comment by pietrodn** [comment]
This is a breaking issue for paying Google Cloud customers, and I guess many more developers using the libraries that depend on gRPC. It prevents developers working on Apple Silicon machines to install their dependencies correctly and get a working development environment.

An immediate action to alleviate the pain would be to "yank" the corrupted M1 wheel on PyPI, and avoid uploading it in future releases. That way, users could compile the library themselves.

With the CircleCI script I shared above, it is possible to build proper Apple Silicon binary wheels, even from Linux. Given the severity of this issue for the developer experience, it would be a huge improvement even to build and upload those wheels by hand until proper automation is available.

**Comment by noblige** [comment]
> > Encountered the same error here with Python 3.10 on Macbook M1. Quick solution for users is to force pip to build the wheel: `pip install --no-binary :all: grpcio --ignore-installed`
> 
> Thanks. This enables me to resolve the previous error message. However, I am seeing the following error now:
> 
> ```
> File "/Users/xin/Program/Pinecone/iac/envaas/./dev.py", line 7, in <module>
>     from pulumi import automation as pa
>   File "/Users/xin/miniconda3/envs/iac_py3/lib/python3.10/site-packages/pulumi/__init__.py", line 33, in <module>
>     from .config import (
>   File "/Users/xin/miniconda3/envs/iac_py3/lib/python3.10/site-packages/pulumi/config.py", line 22, in <module>
>     from . import log
>   File "/Users/xin/miniconda3/envs/iac_py3/lib/python3.10/site-packages/pulumi/log.py", line 22, in <module>
>     from .runtime.settings import get_engine
>   File "/Users/xin/miniconda3/envs/iac_py3/lib/python3.10/site-packages/pulumi/runtime/__init__.py", line 29, in <module>
>     from .mocks import (
>   File "/Users/xin/miniconda3/envs/iac_py3/lib/python3.10/site-packages/pulumi/runtime/mocks.py", line 24, in <module>
>     from . import rpc, rpc_manager
>   File "/Users/xin/miniconda3/envs/iac_py3/lib/python3.10/site-packages/pulumi/runtime/rpc.py", line 29, in <module>
>     from . import known_types, settings
>   File "/Users/xin/miniconda3/envs/iac_py3/lib/python3.10/site-packages/pulumi/runtime/settings.py", line 22, in <module>
>     import grpc
>   File "/Users/xin/miniconda3/envs/iac_py3/lib/python3.10/site-packages/grpc/__init__.py", line 22, in <module>
>     from grpc import _compression
>   File "/Users/xin/miniconda3/envs/iac_py3/lib/python3.10/site-packages/grpc/_compression.py", line 15, in <module>
>     from grpc._cython import cygrpc
> ImportError: dlopen(/Users/xin/miniconda3/envs/iac_py3/lib/python3.10/site-packages/grpc/_cython/cygrpc.cpython-310-darwin.so, 0x0002): symbol not found in flat namespace '_CFRelease'
> ```
> 
> Tried with both 1.43.0 and 1.41.1 version grpc, got the same above error.

I'm seeing the same. Is there a solution?

**Comment by ajatkj** [comment]
@noblige [this stackoverflow answer](https://stackoverflow.com/a/39429110) might give you some insight into what's going on.

**Comment by leuduan** [comment]
I got the same error: `ImportError: ... grpc/_cython/cygrpc.cpython-310-darwin.so, 0x0002): symbol not found in flat namespace '_CFRelease'` when import grpc
After some diggings, the _CFRelease seems to be a symbol of apple foundation framework and not yet loaded into python interpreter. So just need to load that shared library into python interpreter before importing grpc. These are a few ways work for me:

- preload the foundation dynamic shared library into python by setting environment variable `export DYLD_INSERT_LIBRARIES=/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation`' and the start the python interpreter
- load directly in python interpreter:
```python
import ctypes
ctypes.cdll.LoadLibrary('/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation')
import grpc
```
- or import some other libraries that probably load foundation framework for us such as `pandas`, `psutil` before importing grpc

**Comment by hx-markterry** [comment]
I added this to my install script after my usual pip install command:

```
if [[ `uname -s` == "Darwin" && `uname -m` == "arm64" ]]; then
  echo "Reinstalling GRPCIO without binary support for Mac M1..."
  # Reinstall grpcio but this time ignore the wheel as it is not compatible with Mac M1
  GRPCIO_VERSION=`pip list | grep "grpcio " | cut -w -f 2`
  pip uninstall grpcio -y
  pip install --no-binary :all: grpcio==${GRPCIO_VERSION} --ignore-installed
fi
```

**Comment by aronchick** [comment]
just wanted to +1 @hx-markterry's steps worked perfectly for me

**Comment by gnossen** [comment]
[`1.47.0rc1` has been released to PyPi.](https://pypi.org/project/grpcio/1.47.0rc1/). This does not yet include native arm64 wheels, but it does remove the erroneous `universal2` labelled wheels. This will allow you to fall back to a from-source build when using an M1 machine without Rosetta. You can consume it with `pip install grpcio --pre`.

If everything goes well, you should see this in `1.47.0` in a few days.

**Comment by pietrodn** [comment]
I confirm this is solved. `pip install grpcio==1.47.0rc1` on my M1 Pro MacBook Pro now downloads the source distribution, and not the broken wheel. Thank you!

**Comment by gnossen** [comment]
[`1.47.0` is now released to PyPi](https://pypi.org/project/grpcio/1.47.0/). This fixes this particular issue. Stay tuned for further work on Mac OS arm64 support.

**Comment by abax1** [comment]
Any updates on the fix to support apple silicon?  This is still not working for me.

**Comment by pietrodn** [comment]
@abax1 there is no fix to support Apple Silicon AFAIK. If you're interested I made [this unofficial repo](https://github.com/pietrodn/grpcio-mac-arm-build) with Apple Silicon builds (look in Releases)

**Comment by wildone** [comment]
Had to downgrade to 1.44.0 as current version 1.54.2 does not have a wheel for Bug Sur 11.7.4 (20G1120).

**Comment by peggyxia-alj** [comment]
> [`1.47.0` is now released to PyPi](https://pypi.org/project/grpcio/1.47.0/). This fixes this particular issue. Sorry for the long wait on this. Stay tuned for further work on Mac OS arm64 support.

Tried on grpc-1.55.1. This issue is still there on Mac M1 chips with Bazel build.

**Comment by gnossen** [comment]
@peggyxia-alj The Bazel build is not universal2 -- the artifacts labelled `universal2` on pypi are. However, Bazel is a native local build, so if you  build on an arm64 machine, the artifacts should be compatible with arm64.

In general, I would not recommend using our Bazel build to produce artifacts to run anywhere but on the machine which did the build.

**Comment by abasu0713** [comment]
> Encountered the same error here with Python 3.10 on Macbook M1.
> Quick solution for users is to force pip to build the wheel: `pip install --no-binary :all: grpcio --ignore-installed`

Best response in all of internet related to this issue! Concise. Uses basic understanding of how pip wheels work! It's just elegant! Thank you so much!
